src/clients/MultiCallerClient.ts

```diff
@@ -49,28 +49,164 @@ export const unknownRevertReasonMethodsToIgnore = new Set([
+    return nTxns;
+  }
+
+  clearTransactionQueue(chainId: number = null) {
+    if (this.newMulticaller) {
+      if (chainId) {
+        this.txns[chainId] = [];
+        this.valueTxns[chainId] = [];
+      } else {
+        this.txns = {};
+        this.valueTxns = {};
+      }
+    } else this.transactions = [];
+  }
+
+  async executeTransactionQueue(simulate = false): Promise<string[]> {
+    if (this.newMulticaller) {
+      // For compatibility with the existing implementation, flatten all txn hashes into a single array.
+      // To be resolved once the legacy implementation is removed and the callers have been updated.
+      const txnHashes: { [chainId: number]: string[] } = await this.executeTxnQueues(simulate);
+      return Array.from(Object.values(txnHashes)).flat();
+    }
+
+    return this.executeTxnQueueLegacy(simulate);
+  }
+
+  // For each chain, collate the enqueued transactions and process them in parallel.
+  async executeTxnQueues(simulate = false): Promise<{ [chainId: number]: string[] }> {
+    // @todo: This feels like a bodge...Find a better way to produce the set of chainIds that have enqueued txns.
+    const chainIds = [...new Set(Object.keys(this.valueTxns).concat(Object.keys(this.txns)))];
+
+    // One promise per chain for parallel execution.
+    const results = await Promise.allSettled(
+      chainIds
+        .filter((_chainId) => {
+          const chainId = Number(_chainId);
+          return (this.txns[chainId] ?? []).length > 0 || (this.valueTxns[chainId] ?? []).length > 0;
+        })
+        .map((_chainId) => {
+          const chainId = Number(_chainId);
+          return this.executeChainTxnQueue(chainId, simulate);
+        })
+    );
+
+    // Collate the results for each chain.
+    const txnHashes: { [chainId: number]: string[] } = Object.fromEntries(
+      results.map((result, idx) => {
+        const chainId = chainIds[idx];
+        switch (result.status) {
+          case "fulfilled":
+            return [chainId, result.value.map((txnResponse) => txnResponse.hash)];
+          case "rejected":
+            return [chainId, result.reason];
+          default: // Should never occur; log if it ever does.
+            this.logger.warn({
+              at: "MultiCallerClient#executeTxnQueues",
+              message: "Unexpected transaction queue resulting status.",
+              result,
+            });
+            return [chainId, "Unknown error"];
+        }
+      })
+    );
+
+    return txnHashes;
   }
 
-  clearTransactionQueue() {
-    this.transactions = [];
+  // For a single chain, simulate all potential multicall txns and group the ones that pass into multicall bundles.
+  // Then, submit a concatenated list of value txns + multicall bundles. Flush the existing queues on completion.
+  async executeChainTxnQueue(chainId: number, simulate = false): Promise<TransactionResponse[]> {
+    const multicallTxns: AugmentedTransaction[] =
+      this.txns[chainId].length > 0
```

Comment:
Should we make sure this isn't undefined?
```suggestion
      this.txns[chainId]?.length > 0
```

Reply:
Nice catch; fixed here: https://github.com/across-protocol/relayer-v2/pull/424/commits/98a66acb394fde49caa7e71ced54237fcbabef87